src/lib.rs

```diff
@@ -65,124 +67,147 @@ pub fn program_from_file(bin_path: &str) -> Vec<U256> {
+    let program_code = program_from_file(bin_path)?;
     vm.load_program(program_code);
     run(vm, tracers)
 }
 
-pub fn run(mut vm: VMState, tracers: &mut [Box<&mut dyn Tracer>]) -> (U256, VMState) {
+pub fn run(
+    mut vm: VMState,
+    tracers: &mut [Box<&mut dyn Tracer>],
+) -> Result<(U256, VMState), EraVmError> {
     let opcode_table = synthesize_opcode_decoding_tables(11, ISAVersion(2));
     loop {
-        let opcode = vm.get_opcode(&opcode_table);
+        let opcode = vm.get_opcode(&opcode_table)?;
         for tracer in tracers.iter_mut() {
             tracer.before_execution(&opcode, &vm);
         }
-        let gas_underflows = vm.decrease_gas(&opcode);
+        let gas_underflows = vm.decrease_gas(&opcode)?;
 
         if vm.predicate_holds(&opcode.predicate) {
             match opcode.variant {
                 // TODO: Properly handle what happens
                 // when the VM runs out of ergs/gas.
                 _ if vm.running_contexts.len() == 1
-                    && vm.current_context().near_call_frames.is_empty()
+                    && vm.current_context()?.near_call_frames.is_empty()
                     && gas_underflows =>
                 {
                     break
                 }
                 _ if gas_underflows => {
-                    _revert_out_of_gas(&mut vm);
+                    _revert_out_of_gas(&mut vm)?;
                 }
                 Variant::Invalid(_) => todo!(),
                 Variant::Nop(_) => todo!(),
                 Variant::Add(_) => {
-                    _add(&mut vm, &opcode);
+                    _add(&mut vm, &opcode)?;
                 }
-                Variant::Sub(_) => _sub(&mut vm, &opcode),
-                Variant::Jump(_) => _jump(&mut vm, &opcode),
-                Variant::Mul(_) => _mul(&mut vm, &opcode),
-                Variant::Div(_) => _div(&mut vm, &opcode),
+                Variant::Sub(_) => _sub(&mut vm, &opcode)?,
+                Variant::Jump(_) => _jump(&mut vm, &opcode)?,
+                Variant::Mul(_) => _mul(&mut vm, &opcode)?,
+                Variant::Div(_) => _div(&mut vm, &opcode)?,
                 Variant::Context(_) => todo!(),
                 Variant::Shift(_) => match opcode.variant {
-                    Variant::Shift(ShiftOpcode::Shl) => _shl(&mut vm, &opcode),
-                    Variant::Shift(ShiftOpcode::Shr) => _shr(&mut vm, &opcode),
-                    Variant::Shift(ShiftOpcode::Rol) => _rol(&mut vm, &opcode),
-                    Variant::Shift(ShiftOpcode::Ror) => _ror(&mut vm, &opcode),
+                    Variant::Shift(ShiftOpcode::Shl) => _shl(&mut vm, &opcode)?,
+                    Variant::Shift(ShiftOpcode::Shr) => _shr(&mut vm, &opcode)?,
+                    Variant::Shift(ShiftOpcode::Rol) => _rol(&mut vm, &opcode)?,
+                    Variant::Shift(ShiftOpcode::Ror) => _ror(&mut vm, &opcode)?,
                     _ => unreachable!(),
                 },
```

Comment:
We could remove the `unreachable!()` by taking the shift variant from the outer `match`:
```rust
Variant::Shift(shift_variant) => match shift_variant {
    ShiftOpcode::Shl => _shl(&mut vm, &opcode)?,
    ShiftOpcode::Shr => _shr(&mut vm, &opcode)?,
    ShiftOpcode::Rol => _rol(&mut vm, &opcode)?,
    ShiftOpcode::Ror => _ror(&mut vm, &opcode)?,
},
```
This works because it makes the inner `match` exhaustive. And, while the original code is correct, this way we can get help from the compiler if we miss one of the inner arms, due to the removal of the catch-all `_` arm.

Reply:
Correct, it has been applied in one of the latests commits.